projects/openapi-utilities/src/openapi3/types.ts

```diff
@@ -0,0 +1,110 @@
+import { OpenAPIV3 } from 'openapi-types';
+
+export type V3FactType =
+  | 'specification'
+  | 'operation'
+  | 'request-header'
+  | 'request-query'
+  | 'request-cookie'
+  | 'request-path'
+  | 'body'
+  | 'requestBody'
+  | 'field'
+  | 'response'
+  | 'response-header'
+  | 'body-example'
+  | 'component-schema-example';
+
+type _OpenApiV3TraverserFact<T extends V3FactType> = {
+  location: {
+    jsonPath: string;
+  };
+  type: T;
+};
+
+export type OpenApiV3TraverserFact<T extends V3FactType> = T extends T
```

Comment:
I needed this indirection / TS magic to get `if (fact.type === '...') -> narrowed fact`